src/Pact/Analyze/Eval/Term.hs

```diff
@@ -611,6 +611,90 @@ evalTerm = \case
       Some (SObject _) _ -> throwErrorNoLoc "We can't yet analyze calls to `hash` on objects"
       Some _ _           -> throwErrorNoLoc "We can't yet analyze calls to `hash` on non-{string,integer,bool}"
 
+  Pact steps -> do
+    -- We execute through all the steps once (via a left fold), then we execute
+    -- all the rollbacks (via for), in reverse order.
+
+    (rollbacks, _) <- foldlM
+      (\(rollbacks, successChoice)
+        --          ^ do we make it to this step?
+        (Step (tm :< ty) successPath {- s_n -} mEntity mCancelVid mRollback) -> 
+        withSing ty $ do
+
+        -- The first step has no cancel var. All other steps do.
+        cancel <- case mCancelVid of
+
+          -- If this is the first step we just evaluate this term
+          Nothing -> do
+            void $ evalTermWithEntity mEntity tm
+            pure sFalse
+
+          -- ... otherwise, we nondeterministically either succeed or cancel
+          Just (cancelPath {- c_n -}, cancelVid) -> withReset $ do
+            cancel <- view (aeNondets . at cancelVid)
+              ??? "couldn't find cancel var"
+            
+            tagSubpathStart successPath $ sansProv successChoice
+            tagSubpathStart cancelPath $ sansProv $ successChoice .&& cancel
+
+            ite cancel
+              (pure ())
+              (void $ evalTermWithEntity mEntity tm)
+            pure cancel
+
+        let rollbacks' = case mRollback of
+              Nothing       
+                -> rollbacks
+              Just (rollbackPath {- r_n -}, rollback) 
+                -> (rollbackPath, successChoice .&& cancel, rollback):rollbacks
+                --                ^ should this rollback execute?
+        
+        pure (rollbacks', successChoice .&& sNot cancel))
+      ([], sTrue)
+      steps
+
+    for_ rollbacks $ \(path {- r_n -}, execMe, rollback) -> do
+      tagSubpathStart path $ sansProv execMe
+      ite execMe (void $ withReset $ evalETerm rollback) (pure ())
```

Comment:
need to handle the case where we cancel after e.g. step 2 instead of running step 3. and then we need to run *all* rollbacks up to this point